client/engine/engine.go

```diff
@@ -0,0 +1,135 @@
+// Package engine contains the types and imperative code for the business logic of a go-nitro Client.
+package engine // import "github.com/statechannels/go-nitro/client/engine"
+
+import (
+	"math/big"
+
+	"github.com/statechannels/go-nitro/channel/state"
+	"github.com/statechannels/go-nitro/client/engine/store"
+	"github.com/statechannels/go-nitro/protocols"
+	"github.com/statechannels/go-nitro/types"
+)
+
+// Engine is the imperative part of the core business logic of a go-nitro Client
+type Engine struct {
+	// inbound go channels
+	API   chan APIEvent
+	Chain chan ChainEvent
+	Inbox chan Message
+
+	store store.Store // A Store foe persisting important data
+}
+
+// APIEvent is an internal representation of an API call
+type APIEvent struct {
+	ObjectiveToSpawn   protocols.Objective   // try this first
+	ObjectiveToReject  protocols.ObjectiveId // then this
+	ObjectiveToApprove protocols.ObjectiveId // then this
+
+	Response chan Response
+}
+
+// ChainEvent is an internal representation of a blockchain event
+type ChainEvent struct {
+	ChannelId          types.Bytes32
+	Holdings           map[types.Address]big.Int // indexed by asset
+	AdjudicationStatus protocols.AdjudicationStatus
+}
+
+// Message is an internal representation of a message from another client
+type Message struct {
+	ObjectiveId protocols.ObjectiveId
+	Sigs        map[types.Bytes32]state.Signature // mapping from state hash to signature
+}
+
+// Response is the return type that asynchronous API calls "resolve to". Such a call returns a go channel of type Response.
+type Response struct{}
+
+// NewEngine is the constructor for an Engine
+func New() Engine {
+	e := Engine{}
+
+	// create the engine's inbound channels
+	e.API = make(chan APIEvent)
+	e.Chain = make(chan ChainEvent)
+	e.Inbox = make(chan Message)
+
+	return e
+}
+
+// Run kicks of an infinite loop that waits for communications on the supplied channels, and handles them accordingly
+func (e *Engine) Run() {
+	for {
+		select {
+		case apiEvent := <-e.API:
+			e.handleAPIEvent(apiEvent)
+
+		case chainEvent := <-e.Chain:
+			e.handleChainEvent(chainEvent)
+
+		case message := <-e.Inbox:
+			e.handleMessage(message)
+
+		}
+	}
+}
+
+// handleMessage handles a Message from a peer go-nitro Wallet.
+// It
+// reads an objective from the store,
+// generates an updated objective and declaration of side effects,
+// commits the updated objective to the store,
+// executes the side effects and
+// evaluates objecive progress.
+func (e *Engine) handleMessage(message Message) {
+	objective := e.store.GetObjectiveById(message.ObjectiveId)
+	event := protocols.ObjectiveEvent{Sigs: message.Sigs}
+	updatedProtocol := objective.Update(event)
+	e.store.SetObjective(updatedProtocol)
```

Comment:
We could certainly do it that way. Personally, I like the conceptual separation between updating the extended state of the protocol with external information (on the one hand) and actively making progress / generating side effects (on the other). The one is passive and the other is active, and encourages (forces, even) objective implementations to decouple those two pathways. To me that is important, since an objective's first responsibilty is to store the information it recieves -- and only then to start signing things and taking action.

We can easily add an `updateAndCrank()` method to objectives that simply calls `update()` and then `crank()`. (We could also call that wrapper `crank()` and have the other methods called `update()` and `progress()` or something).

I suppose the obvious question is, do we _need_ the ability to do things in a more general way? For example, to sometimes update and then progress, and sometimes progress then update. I think if the answer is no, this separation is likely to lead to cleaner code because it imposes more restrictions on objective implementations. 

BTW I've tried to do justice to these ideas in the `Objective` interface itself.

Reply:
> I like the conceptual separation between updating the extended state of the protocol with external information (on the one hand) and actively making progress / generating side effects (on the other).

I strongly agree. But actively making progress may also need to update the objective's state. For instance, it may record a timestamp or an idempotency key.

In this case, the seperation is around:
- updating the objective's state as a consequence of processing an event
- making progress based on the current state

The point here is that both `update(someEvent)` and `crank()` may wish to update the objective's state.